### PR TITLE
Implement einsum

### LINKED ISSUE #2665: Add Tensor.einsum
Then [upstream it to einops](https://github.com/arogozhnikov/einops/pull/297), by overriding [this method](https://github.com/arogozhnikov/einops/blob/fcd36c9b017d52e452a301e75c1373b75ec23ee0/einops/_backends.py#L133).

### PR DESCRIPTION
**Pull request by blueridanus**
Closes #2665. Is this desired in core? Seems appropriate to me, but it doesn't take self, so if LOC is a problem, could be in extra.

- [x] untested einsum impl
- [x] add a few simple tests
- [x] cover every case [here](https://rockt.github.io/2018/04/30/einsum)
- [x] make sure transpose shenanigans are indeed covered by tests
- [x] add mustfails for bad syntax, mismatched shapes
- [x] check if it works in einops
- [x] add more helpful error msg on dim mismatch
- [x] perf comparisons

### CODE REVIEW

**Comment by blueridanus** [comment]
Would be neat to see how the einsum version of e.g. matmul compares performance wise.

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -510,6 +510,31 @@ def argmax(self, axis=None, keepdim=False):
     return self.shape[axis]-idx.max(axis=axis, keepdim=keepdim)-1
   def argmin(self, axis=None, keepdim=False): return (-self).argmax(axis=axis, keepdim=keepdim)
 
+  @staticmethod
+  def einsum(formula, *xs) -> Tensor:
+    if isinstance(xs[0], list): xs = tuple(xs[0])
```
Comment (geohot) [inline tinygrad/tensor.py:515]: Can fixarg do this?

**Comment by geohot** [comment]
Yea it should be in core, as beautiful as possible.

**Comment by geohot** [comment]
> Would be neat to see how the einsum version of e.g. matmul compares performance wise.

Perf will be identical, test it :)

File: tinygrad/tensor.py (hunk 1 above)
Comment (blueridanus) [reply tinygrad/tensor.py]: yep, nice

**Comment by blueridanus** [comment]
Works in einops:
image

**Comment by blueridanus** [comment]
> > Would be neat to see how the einsum version of e.g. matmul compares performance wise.
> 
> Perf will be identical, test it :)

Seems pretty close:
image

**Comment by geohot** [comment]
Cool!

**Pull request merged**